uber/config.py

```diff
@@ -72,6 +72,65 @@ def PREREG_BADGE_TYPES(self):
         return types
 
     @property
+    def BADGE_DISPLAY_TYPES(self):
+        """
+        There are several contexts where we want to display different badge types to select:
+        - A new attendee registering for the event
+        - An attendee claiming a blank badge in an existing group
+        - An existing attendee editing their registration
+
+        This property uses the parameters in the URL to figure out the context and return the correct
+        badge types to display for an attendee.
+        """
+        with sa.Session() as session:
+            params = cherrypy.lib.httputil.parse_query_string(cherrypy.request.query_string)
+            group = session.group(params['group_id']) if 'group_id' in params else None
+            attendee = session.attendee(params['id']) if 'id' in params else None
+
+            if not attendee:
+                # Inherits the group's badge type if it's an attendee claiming a badge in a group
+                base_badge_name = group.ribbon_and_or_badge if group else c.BADGES.get(c.ATTENDEE_BADGE)
+            else:
+                base_badge_name = attendee.ribbon_and_or_badge
+
+            donation_prepend = '' if base_badge_name == c.BADGES.get(c.ATTENDEE_BADGE) else base_badge_name + ' / '
+            badge_cost = attendee.badge_cost if attendee else c.BADGE_PRICE
+
+            # The base badge and the group badge are special, so they're given default values and added manually
+            base_description = 'Allows access to the convention for its duration.'
+
+            if c.BEFORE_GROUP_PREREG_TAKEDOWN:
+                group_description = 'Register a group of ' + str(c.MIN_GROUP_SIZE) + ' or more and save $' + \
+                                    str(c.GROUP_DISCOUNT) + ' per badge.'
+            else:
+                group_description = 'The deadline for Group registration has passed, but you can still register as ' + \
+                                    'a regular attendee.'
+
+            badge_types = {}
+            badge_types['base'] = {
+                'title': base_badge_name + ': $' + str(badge_cost),
+                'description': base_description,
+                'extra': 0
+            }
+
+            if c.GROUPS_ENABLED and not attendee and not group:
+                badge_types['group'] = {
+                    'title': 'Group Leader',
+                    'description': group_description,
+                    'extra': 0
+                }
+
+            # The rest of the badges are added in via config
+            for name, option in c.BADGE_DISPLAY_CONFIGS.items():
+                if int(option['extra']) in c.PREREG_DONATION_TIERS:
```

Comment:
This function is missing the ability to check the availability of certain levels. I haven't yet decided the best way to support configuring this, since the _AVAILABLE properties themselves are dynamic (and, specifically, aren't guaranteed to exist).